Add process instrumentation

danihodovic opened the pull request:
# Checklist:

- [ ] Code follows the style guidelines
- [x] ~Documentation updated~
- [ ] Corner cases evaluated and handled
- [x] Automatic tests added/update
- [x] Automatic tests pass
- [x] Software runs (locally or hosted)
- [x] Sanity check passed
- [x] Self-review passed

# Changes

I've seen unhandledRejection errors bubble up to the process when
operating Hermes workers in production. This causes a zombie process
which isn't operational, but continues running.

The commit introduces the following changes:

on unhandledRejection or uncaughtException errors that are not handled
and bubble up to the process, do the following:

- exit with a status code of 1 so that the process is restarted by it's
  supervisor in a fresh state (in our case by the Docker daemon)

- log the error

- report the error to Sentry (if Sentry is enabled)

Additionally we enable Sentry for the Hermes and Atlas workers rather
than having it run only on the server.

danihodovic commented:
Rebase on #322

spherefoundry commented on src/utils/setup_process.js:
```diff
@@ -0,0 +1,40 @@
+import config from '../../config/config';
+
+const reportAndExit = (logger, message, error) => {
+  Sentry.captureException(error);
```
I would assume this needs some time to send the error msg to Sentry. Is this synchronous? If not, the request will never go through because of the process.exit(1) just 2 lines below.

conorseabrook replied:
It is not synchronous, but there is a way to flush the buffer. I saw an example in the docs on how to flush the queue and then exit on the completion callback.

danihodovic replied:
@conorseabrook do you have a link?

danihodovic replied:
Found it https://docs.sentry.io/error-reporting/configuration/draining/?platform=javascript thanks ;)

conorseabrook replied:
👍 I was on my phone or I would have posted.

danihodovic commented:
I've added tests and added logic for flushing Sentry events.

There is a test missing for `uncaughtExceptions` because Mocha makes this really hard to test. See:
https://stackoverflow.com/questions/9025095/how-can-i-test-uncaught-errors-in-mocha
https://stackoverflow.com/questions/18786325/mocha-unit-test-returns-0-as-exit-code-on-uncaught-exceptions
https://gist.github.com/othiym23/4179636

I've tried `prependOnceListener` and removing listeners for uncaughtException and didn't manage to make it work.

If anyone has any ideas or workarounds please let me know.
